tools: add unescaped regexp dot rule to linter

mscdex opened the pull request:
This adds a custom rule to the linter that checks for unescaped "literal" dots in regular expressions. For regexp literals this is straightforward, but it's trickier for `RegExp()` usage because of possible variable text passed to the constructor. This rule tries its best in the constructor scenario by concatenating any/all of the string literals passed to it and checking that.

CI: https://ci.nodejs.org/job/node-test-pull-request/6825/

##### Checklist

- [X] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [X] commit message follows commit guidelines

##### Affected core subsystem(s)

* tools
* test

mscdex commented:
/cc @nodejs/testing

not-an-aardvark commented on tools/eslint-rules/unescaped-regexp-dot.js:
```diff
@@ -0,0 +1,119 @@
+const utilsPath = path.join(__dirname, '..', 'eslint', 'lib', 'ast-utils.js');
+const astUtils = require(utilsPath);
+const getLocationFromRangeIndex = astUtils.getLocationFromRangeIndex;
+const getRangeIndexFromLocation = astUtils.getRangeIndexFromLocation;
```
These methods aren't part of the public API, and were removed in v3.17.0. (Using them for now is probably fine, but they'll need to be changed to `sourceCode.getLocFromIndex` and `sourceCode.getIndexFromLoc` when ESLint is upgraded.)

mscdex replied:
Yes, I know.

mscdex replied:
FWIW I've added a check to support the later versions too.

not-an-aardvark commented on tools/eslint-rules/unescaped-regexp-dot.js:
```diff
@@ -0,0 +1,119 @@
+            break;
+          case '.':
+            if (!escaping) {
+              if (!inCharClass && (i + 1) < str.length) {
```
Would we want to report the node when `i + 1 === str.length`? This would cover the case where the `.` is at the end of the regex, e.g.

```js
assert.throws(func, /this error message ends with a dot./);
```

mscdex replied:
Changed.

not-an-aardvark commented on tools/eslint-rules/unescaped-regexp-dot.js:
```diff
@@ -0,0 +1,119 @@
+  return {
+    Literal: function(node) {
+      if (inRegExp && typeof node.value === 'string' && node.value.length)
+        regexpBuffer.push([node, node.value]);
```
This could cause some weird behavior if the argument to `RegExp` isn't just string concatenation:

```js
// 'foo.bar' gets interpreted as a regex pattern
new RegExp(function() {
  return crypto.createHash('sha1').update('foo.bar').digest('hex');
}());
```

Or it could handle character classes incorrectly:

```js
const closingBracket = ']';

new RegExp('[abc' + closingBracket + '.');
```

There's no perfect solution to this, so it's probably fine as-is if we're not running into cases like this.

mscdex replied:
Regarding the first example, I originally had it coded differently and actually checked parent nodes when encountering string literals, but that seemed more complicated.

Regarding the second example, I pointed that out in the PR description. I doubt that would ever become a problem in practice though.

mscdex replied:
I've re-added the strict parent node checking.

mscdex commented:
I've addressed raised concerns and added support for template literals. These new changes resulted in several more instances being found.

CI: https://ci.nodejs.org/job/node-test-pull-request/6832/

fhinkel commented:
I think this needs a rebase.

fhinkel commented:
ping @mscdex

mscdex commented:
@fhinkel Rebased.

mscdex commented:
CI again: https://ci.nodejs.org/job/node-test-linter/7893/

fhinkel commented:
Thanks. Landed in 61ebfa8d1fcb468cb95fdf3742e1672e48672b32

gibfahn commented:
Are you willing to backport this to `v6.x-staging`? If yes please follow [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md) and raise a backport PR, if no let me know or add the `dont-land-on` label.

MylesBorins commented:
ping one more time re: backport
